### DISCUSSION TITLE
Fprime import explicit cpp implementation in fpp files

### DISCUSSION THREAD

**Discussion by DentOpt** [nasa/fprime#1631] [Q&A] [answered]
Hi, 

I am having a type which is defined in
```
Types/MyTypeSerializable.cpp
Types/MyTypeSerializable.cpp
```
Now I would like to use this type in a port which is defined in e.g. MyModule/MyComponentApp.fpp
```
module MyModule {

    queued component MyComponent {
         ...
         sync input port GetMyType: [1] Types.MyType
         ...
    }
}
```
How can I import such explicitly in cpp defined Types in fpp files, something alike
```
  include type "Types::MyType" from "../Types/MyType.hpp"
```
?
Thank you in advance

**Comment by LeStarch** [comment, marked as answer]
It looks like this is a C++ type that you want to use in an FPP model (as opposed to a FPP defined type for use in the FPP model).  @bocchino is this possible?  Or does the type need to be defined in FPP if it is to be used in an FPP model?

**Reply by codeflight1** [reply]
This is possible simply by defining it as `type MyType`. See https://fprime-community.github.io/fpp/fpp-users-guide.html#Defining-Types_Abstract-Type-Definitions for more info.

**Reply by bocchino** [reply]
This answer is correct. See also https://fprime-community.github.io/fpp/fpp-users-guide.html#Writing-C-Plus-Plus-Implementations_Implementing-Abstract-Types for a discussion of the interface contract between the C++ type and the FPP-generated code.

**Reply by DentOpt** [reply]
Hi,

thank you already for your answers!
In general this approach works, but is not compatible with the previous file naming convention 

In the fpp I can use the abstract type
```
type MyTpe
```
Then it will generate include paths:
```
<include_header>Types/MyType.hpp</include_header>
```

So in order to make this work, I have to change the file names to 
```
Types/MyType.hpp
Types/MyType.cpp
```
This is working. Is there any way on specifying the include path for the abstract type ?
```
<include_header>Types/MyTypeSerializable.hpp</include_header>
```

**Reply by bocchino** [reply]
If you want the include path to be Foo/Bar/MyType.hpp, then you can define the type MyType in a model file Foo/Bar/MyType.fpp. In general the location of the header path for an FPP type T is the location of the model file where T is defined.

**Reply by DentOpt** [reply]
Hi, for the path that's clear, but does not solve the issue with the file name `Foo/Bar/MyTypeSerializable.hpp`. 
It would be good to have an option to be able to specify the include hpp, e.g.
```
type MyType at "MyComponent/MyTypeSerializable.hpp"
```

**Reply by bocchino** [reply]
Yes, we could add that feature. I'll add it to the list of proposed new features. For now you can rename the file.

**Reply by DentOpt** [reply]
Excellent, thank you for your support!

**Comment by mohamed-ghita** [comment]
Hello, What is the state on this? 
@bocchino Do you have reference to the feature, please?

**Reply by bocchino** [reply]
It's in the list of proposed FPP features: https://github.com/fprime-community/fpp/issues/163.
